grammar.js

```diff
@@ -446,15 +451,59 @@ module.exports = grammar({
     _float_literal: $ => choice($._hex_literal, $._number_literal),
 
     // Strings.
-    // Regex is: "Between two quote characters, allow either backslash followed by any
-    // character (i.e. escape sequence) or any character except that quote character or
-    // a bare backtick."
     string: $ => choice(
       $._raw_string_literal,
-      /'((?:\\(.|\n))|[^'\\])*'/,
-      /"((?:\\(.|\n))|[^"\\])*"/,
+      $._single_quoted_string,
+      $._double_quoted_string
+    ),
+
+    // TODO: Raw string contents, something like this, where `_raw_string_open`,
+    // `_raw_string_close`, and `_raw_string_content` are externals.
+    // _raw_string_literal: $ => seq(
+    //   $._raw_string_open,
+    //   optional(field("content", alias($._raw_string_content, $.string_content))),
+    //   $._raw_string_close
+    // ),
+
+    // Explanation is:
+    // - Between two quote characters, allow either:
+    //   - Anything except `'` (or `"`) or `\`
+    //   - An escape sequence
+    _single_quoted_string: $ => seq(
+      '\'',
+      optional(field("content", alias($._single_quoted_string_content, $.string_content))),
+      '\''
     ),
 
+    _double_quoted_string: $ => seq(
+      '"',
+      optional(field("content", alias($._double_quoted_string_content, $.string_content))),
+      '"'
+    ),
+
+    _single_quoted_string_content: $ => repeat1(choice(
+      /[^'\\]+/,
+      $.escape_sequence
+    )),
+
+    _double_quoted_string_content: $ => repeat1(choice(
+      /[^"\\]+/,
+      $.escape_sequence
+    )),
+
+    escape_sequence: $ => token.immediate(seq(
+      '\\',
+      choice(
+        /[^0-9xuU]/,
+        /[0-7]{1,3}/,
+        /x[0-9a-fA-F]{1,2}/,
+        /u[0-9a-fA-F]{1,4}/,
+        /u\{[0-9a-fA-F]{1,4}\}/,
+        /U[0-9a-fA-F]{1,8}/,
+        /U\{[0-9a-fA-F]{1,8}\}/
```

Comment:
Notable changes from `main` include:
- `/[^0-9xuU]/` now excluding `\8` and `\9`
- `/u\{[0-9a-fA-F]{1,4}\}/` was `/u\{[0-9a-fA-F]+\}/`, but really only up to 4 values are allowed within the `{}`
- `/U` with the capital U was not supported at all before
- `/x[0-9a-fA-F]{1,2}` was `/x[0-9a-fA-F]{2}`, but `\x1` is valid, so should be `{1,2}` (same for `\u`, which was `{4}` before)